From 25b390440c75b3937d7f22c73675a49c00a42bb3 Mon Sep 17 00:00:00 2001
From: Aikar <aikar@aikar.co>
Date: Thu, 25 Jul 2013 23:08:50 -0400
Subject: [PATCH] Improvements to Player lookup - Resolves BUKKIT-4608

Here is the rest of my changes that someone decided to remove for no reason at all...
---
 .../java/org/bukkit/craftbukkit/CraftServer.java   | 24 +++++++++++++++-------
 1 file changed, 17 insertions(+), 7 deletions(-)

diff --git a/src/main/java/org/bukkit/craftbukkit/CraftServer.java b/src/main/java/org/bukkit/craftbukkit/CraftServer.java
index dd97003..2b010d2 100644
--- a/src/main/java/org/bukkit/craftbukkit/CraftServer.java
+++ b/src/main/java/org/bukkit/craftbukkit/CraftServer.java
@@ -506,12 +506,19 @@ public final class CraftServer implements Server {
     public Player getPlayer(final String name) {
         Validate.notNull(name, "Name cannot be null");
 
-        Player[] players = getOnlinePlayers();
+        // Spigot End
+        Player directLookup = getPlayerExact( name );
+        if ( directLookup != null )
+        {
+            return directLookup;
+        }
+        // Spigot End
 
         Player found = null;
         String lowerName = name.toLowerCase();
         int delta = Integer.MAX_VALUE;
-        for (Player player : players) {
+        for (EntityPlayer entityPlayer : (List<EntityPlayer>) playerList.players) {
+            Player player = entityPlayer.playerConnection.getPlayer();
             if (player.getName().toLowerCase().startsWith(lowerName)) {
                 int curDelta = player.getName().length() - lowerName.length();
                 if (curDelta < delta) {
@@ -562,7 +569,8 @@ public final class CraftServer implements Server {
 
         List<Player> matchedPlayers = new ArrayList<Player>();
 
-        for (Player iterPlayer : this.getOnlinePlayers()) {
+        for (EntityPlayer entityPlayer : (List<EntityPlayer>) playerList.players) {
+            Player iterPlayer = entityPlayer.getBukkitEntity();
             String iterPlayerName = iterPlayer.getName();
 
             if (partialName.equalsIgnoreCase(iterPlayerName)) {
@@ -1528,7 +1536,8 @@ public final class CraftServer implements Server {
     public void sendPluginMessage(Plugin source, String channel, byte[] message) {
         StandardMessenger.validatePluginMessage(getMessenger(), source, channel, message);
 
-        for (Player player : getOnlinePlayers()) {
+        for (EntityPlayer entityPlayer : (List<EntityPlayer>) playerList.players) {
+            Player player = entityPlayer.getBukkitEntity();
             player.sendPluginMessage(source, channel, message);
         }
     }
@@ -1536,7 +1545,8 @@ public final class CraftServer implements Server {
     public Set<String> getListeningPluginChannels() {
         Set<String> result = new HashSet<String>();
 
-        for (Player player : getOnlinePlayers()) {
+        for (EntityPlayer entityPlayer : (List<EntityPlayer>) playerList.players) {
+            Player player = entityPlayer.getBukkitEntity();
             result.addAll(player.getListeningPluginChannels());
         }
 
@@ -1641,11 +1651,11 @@ public final class CraftServer implements Server {
     }
 
     public List<String> tabCompleteChat(Player player, String message) {
-        Player[] players = getOnlinePlayers();
         List<String> completions = new ArrayList<String>();
         PlayerChatTabCompleteEvent event = new PlayerChatTabCompleteEvent(player, message, completions);
         String token = event.getLastToken();
-        for (Player p : players) {
+        for (EntityPlayer entityPlayer : (List<EntityPlayer>) playerList.players) {
+            Player p = entityPlayer.getBukkitEntity();
             if (player.canSee(p) && StringUtil.startsWithIgnoreCase(p.getName(), token)) {
                 completions.add(p.getName());
             }
-- 
1.9.1

